File: 算法训练营/树/144/迭代法.py
#
# @lc app=leetcode.cn id=144 lang=python3
#
# [144] 二叉树的前序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        stack = [root]
        records, ret = set(), []
        while stack:
            elem = stack.pop()
            if elem is None:
                continue
            if elem not in records:
                ret.append(elem.val)
                records.add(elem)
            if elem.left and elem.left not in records:
                stack.extend([elem, elem.left])
                continue
            if elem.right:
                stack.append(elem.right)
        return ret

# @lc code=end

File: 算法训练营/20-字符串算法/242/迭代法.go
/*
 * @lc app=leetcode.cn id=242 lang=golang
 *
 * [242] 有效的字母异位词
 */

// @lc code=start
func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	var sArr[26]byte
	var tArr[26]byte
	for i := 0; i < len(s); i++ {
		sArr[s[i]-'a'] += 1
		tArr[t[i]-'a'] += 1
	}

	return sArr == tArr

}
// @lc code=end
